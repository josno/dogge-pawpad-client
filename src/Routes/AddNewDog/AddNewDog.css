.block {
	display: block;
	font-size: 1em;
	height: 40px;
	padding-left: 10px;
	margin-bottom: 10px;
	width: 100%;
}

.bold {
	color: #f7567c;
	font-size: 0.8em;
	font-weight: bold;
}

.fielditem {
	height: 80px;
}

.shot-date-input {
	display: block;
	text-align: center;
	font-size: 0.8em;
	height: 30px;
	margin: 10px auto 10px auto;
	width: 80%;
}

.react-datepicker-wrapper {
	width: 100%;
}

.loader-container {
	width: 100%;
	display: flex;
	justify-content: center;
	position: relative;
	bottom: 50vh;
}

.img-preview {
	width: 250px;
}

.loader {
	border: 16px solid gray;
	border-radius: 50%;
	opacity: 0.5;
	border-top: 16px solid #ff7300;
	width: 120px;
	height: 120px;
	-webkit-animation: spin 2s linear infinite; /* Safari */
	animation: spin 2s linear infinite;
}

/* Safari */
@-webkit-keyframes spin {
	0% {
		-webkit-transform: rotate(0deg);
	}
	100% {
		-webkit-transform: rotate(360deg);
	}
}

@keyframes spin {
	0% {
		transform: rotate(0deg);
	}
	100% {
		transform: rotate(360deg);
	}
}

fieldset {
	display: flex;
	margin: 10px 0px 10px 0px;
}
