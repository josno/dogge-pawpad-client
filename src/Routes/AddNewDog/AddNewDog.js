import React, { Component } from "react";
import PawPadContext from "../../PawPadContext.js";
import { Link } from "react-router-dom";
import DogsApiService from "../../services/api-service";
import "./AddNewDog.css";
import ValidationError from "../../Components/ValidationError/ValidationError";
import Validate from "../../Utils/validation";

import DatePicker from "react-datepicker";

import "react-datepicker/dist/react-datepicker.css";

class AddNewDog extends Component {
	static contextType = PawPadContext;
	constructor(props) {
		super(props);
		this.state = {
			dogName: { value: "", touched: false },
			profileImgPreview: "",
			profileImg: null,
			spayedneutered: false,
			rabies: "",
			rabiesBooster: "",
			complexOne: "",
			complexTwo: "",
			complexBooster: "",
			microchip: null,
			tagNumber: null,
			gender: null,
			loading: false,
			age: { value: null, touched: false },
			arrivalDate: null,
			error: null,
		};
		this.handleImgChange = this.handleImgChange.bind(this);
		this.handleCheckbox = this.handleCheckbox.bind(this);
		this.handleChange = this.handleChange.bind(this);
		this.handleSubmit = this.handleSubmit.bind(this);
		this.setShotsObject = this.setShotsObject.bind(this);
		this.setDogObject = this.setDogObject.bind(this);
		this.setFormData = this.setFormData.bind(this);
		this.setLoading = this.setLoading.bind(this);
	}

	handleImgChange = (e) => {
		this.setState({
			profileImgPreview: URL.createObjectURL(e.target.files[0]),
			profileImg: e.target.files[0],
		});
	};

	handleChange = (e) => {
		const { name, value } = e.target;
		this.setState({
			[name]: value,
		});
	};

	stringifyDate = (date) => {
		const event = new Date(`${date}`);
		return JSON.stringify(event).slice(1, 11);
	};

	handleDateChange = (date) => {
		this.setState({
			arrivalDate: date,
		});
	};

	updateAge(date) {
		this.setState({
			age: { value: date, touched: true },
		});
	}

	updateName(str) {
		const name = str.trim();
		this.setState({
			dogName: { value: name, touched: true },
		});
	}

	handleCheckbox = (e) => {
		const { name } = e.target;

		this.setState((prevState) => ({
			[name]: !prevState[name],
		}));
	};

	setShotsObject = () => {
		const {
			rabies,
			complexOne,
			complexTwo,
			rabiesBooster,
			complexBooster,
		} = this.state;

		const shots = [
			{
				shot_name: "Rabies",
				shot_date: rabies.length > 0 ? rabies : null,
				dog_id: "",
			},
			{
				shot_name: "Rabies Yearly Booster",
				shot_date: rabiesBooster.length > 0 ? rabiesBooster : null,
				dog_id: "",
			},
			{
				shot_name: "Complex I",
				shot_date: complexOne.length > 0 ? complexOne : null,
				dog_id: "",
			},
			{
				shot_name: "Complex II",
				shot_date: complexTwo.length > 0 ? complexTwo : null,
				dog_id: "",
			},
			{
				shot_name: "Complex Yearly Booster",
				shot_date: complexBooster.length > 0 ? complexBooster : null,
				dog_id: "",
			},
		];

		shots.forEach((i) => {
			if (i.shot_date !== null) {
				i.shot_iscompleted = true;
			} else {
				i.shot_iscompleted = false;
			}
		});

		return shots;
	};

	setDogObject = (
		spayedNeutered,
		gender,
		microchip,
		tagNumber,
		arrivalDate,
		age,
		dogName
	) => {
		const arrivalDateString = this.stringifyDate(arrivalDate);
		const ageDateString = this.stringifyDate(age);
		const newDog = [
			{ dog_name: dogName },
			{ spayedneutered: spayedNeutered },
			{ gender: gender },
			{ microchip: microchip },
			{ tag_number: tagNumber },
			{ age: ageDateString },
			{ arrival_date: arrivalDateString },
		];

		return newDog;
	};

	setFormData = (newDog, profileImg) => {
		const formData = new FormData();
		formData.append("profile_img", profileImg);

		newDog.forEach((i) => {
			formData.append(Object.keys(i), Object.values(i));
		});

		return formData;
	};

	setLoading = (e) => {
		this.setState({
			loading: true,
		});
	};

	handleSubmit = (e) => {
		e.preventDefault();
		const {
			profileImg,
			spayedNeutered,
			gender,
			arrivalDate,
			tagNumber,
			microchip,
		} = this.state;
		const dogName = this.state.dogName.value;
		const age = this.state.age.value;

		const newDog = this.setDogObject(
			spayedNeutered,
			gender,
			microchip,
			tagNumber,
			arrivalDate,
			age,
			dogName
		);
		const allCompletedShots = this.setShotsObject();
		const formData = this.setFormData(newDog, profileImg);

		DogsApiService.insertNewDog(formData)
			.then((res) => {
				allCompletedShots.map((i) => (i.dog_id = res.id));
				allCompletedShots.map((shot) => DogsApiService.insertNewShot(shot));
			})
			.then((res) => this.props.history.push("/dogs-list"))
			.catch((error) => {
				this.setState({ error: error.message });
			});
	};

	componentDidMount() {
		window.scrollTo(0, 0);
	}

	render() {
		console.log(this.state.arrivalDate);
		let imgStyle = { display: "none" };

		if (this.state.profileImgPreview.length > 0) {
			imgStyle = {
				display: "block",
			};
		}

		return (
			<main className='add-dog-container'>
				<h1 className='form-title'>Add New Dog</h1>

				<form className='form-container' onSubmit={this.handleSubmit}>
					<div className='field-item'>
						<label htmlFor='name' className='bold'>
							Name
						</label>
						<input
							id='name'
							type='text'
							name='dogName'
							className='block'
							onChange={(e) => this.updateName(e.target.value)}
							required
						/>
						{this.state.dogName.touched && (
							<ValidationError
								message={Validate.validateName(this.state.dogName.value)}
							/>
						)}
					</div>

					<div className='field-item'>
						<label htmlFor='image' className='bold'>
							Image
						</label>
						<input
							className='block'
							type='file'
							name='profileImg'
							onChange={this.handleImgChange}
							accept='image/*'
							required
						/>
						<img
							id='img-preview'
							src={this.state.profileImgPreview}
							alt='your-pic'
							style={imgStyle}
						/>
					</div>

					<fieldset className='field-item'>
						<legend className='bold'>Gender</legend>
						<label htmlFor='male'>
							<input
								type='radio'
								name='gender'
								value='Male'
								onChange={this.handleChange}
								id='male'
								required
							/>
							Male
						</label>

						<label htmlFor='female'>
							<input
								type='radio'
								name='gender'
								value='Female'
								onChange={this.handleChange}
								id='female'
							/>
							Female
						</label>
					</fieldset>

					<label htmlFor='estimated-age' className='bold'>
						Estimated Birthdate
					</label>
					<DatePicker
						dateFormat='dd/MM/yyyy'
						selected={this.state.age.value}
						onChange={(e) => this.updateAge(e)}
						placeholderText='Click to select a date'
						id='arrival'
						className='block'
						showYearDropdown
						dateFormatCalendar='MMMM'
						yearDropdownItemNumber={5}
						scrollableYearDropdown
					/>
					{/* <input
						className='block'
						id='age'
						type='text'
						name='age'
						placeholder='example: 12/14/2019 or 12/xx/2019'
						onChange={(e) => this.updateAge(e.target.value)}
						required
					/> */}
					<label htmlFor='arrival' className='bold'>
						Estimated Arrival
					</label>

					<DatePicker
						dateFormat='dd/MM/yyyy'
						selected={this.state.arrivalDate}
						placeholderText='Click to select a date'
						onChange={(e) => this.handleDateChange(e)}
						id='arrival'
						className='block'
					/>
					<label htmlFor='tag-number' className='bold'>
						Tag Number
					</label>
					<input
						className='block'
						id='tag-number'
						type='text'
						name='tagNumber'
						onChange={this.handleChange}
					/>

					<label htmlFor='microchip' className='bold'>
						Microchip Number
					</label>
					<input
						className='block'
						id='microchip'
						type='text'
						name='microchip'
						onChange={this.handleChange}
					/>

					<fieldset className='field-item'>
						<legend className='bold'>Spayed/Neutered </legend>
						<label htmlFor='yes'>
							<input
								id='yes'
								type='radio'
								name='spayedNeutered'
								onChange={this.handleCheckbox}
								required
							/>
							Yes
						</label>

						<label htmlFor='no'>
							<input
								id='no'
								type='radio'
								name='spayedNeutered'
								onChange={this.handleCheckbox}
							/>
							No
						</label>
					</fieldset>

					<fieldset className='field-item '>
						<legend className='bold'>Required Shots Completed</legend>

						<label htmlFor='rabies'>
							Rabies{" "}
							<input
								className='shot-date-input'
								id='rabies'
								type='date'
								name='rabies'
								value={this.state.rabies}
								onChange={this.handleChange}
							/>
						</label>

						<label htmlFor='rabiesBooster'>
							Rabies Yearly Booster{" "}
							<input
								className='shot-date-input'
								id='rabiesBooster'
								type='date'
								name='rabiesBooster'
								value={this.state.rabiesBooster}
								onChange={this.handleChange}
							/>
						</label>

						<label htmlFor='complexOne'>
							Complex I
							<input
								className='shot-date-input'
								id='complexOne'
								type='date'
								name='complexOne'
								value={this.state.complexOne}
								onChange={this.handleChange}
							/>
						</label>

						<label htmlFor='complexTwo'>
							Complex II
							<input
								className='shot-date-input'
								id='complexTwo'
								type='date'
								name='complexTwo'
								value={this.state.complexTwo}
								onChange={this.handleChange}
							/>
						</label>

						<label htmlFor='complexBooster'>
							Complex Yearly Booster
							<input
								className='shot-date-input'
								id='complexBooster'
								type='date'
								name='complexBooster'
								value={this.state.complexBooster}
								onChange={this.handleChange}
							/>
						</label>
					</fieldset>

					{this.state.error !== null && (
						<ValidationError
							message={"Something wrong happened. Refresh and try again."}
						/>
					)}
					<div className='form-buttons'>
						<button className='cancel'>
							<Link className='dog-link' to={"/dogs-list"}>
								Cancel
							</Link>
						</button>

						<button
							className='submit'
							type='submit'
							onClick={this.setLoading}
							disabled={Validate.validateName(this.state.dogName.value)}
						>
							Submit
						</button>
					</div>
				</form>
				{this.state.loading && (
					<div className='loader-container'>
						<div className='loader'></div>
					</div>
				)}
			</main>
		);
	}
}

export default AddNewDog;
