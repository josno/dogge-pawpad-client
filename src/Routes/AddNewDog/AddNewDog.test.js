import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import AddNewDog from './AddNewDog';

describe('<AddNewDog />', () => {
	it('renders without crashing', () => {
		const div = document.createElement('div');
		ReactDOM.render(
			<BrowserRouter>
				<AddNewDog />
			</BrowserRouter>,
			div
		);
		ReactDOM.unmountComponentAtNode(div);
	});

	it('renders as expected', () => {
		const tree = renderer
			.create(
				<BrowserRouter>
					<AddNewDog />
				</BrowserRouter>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
