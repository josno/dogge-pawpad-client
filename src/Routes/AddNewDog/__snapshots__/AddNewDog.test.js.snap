// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddNewDog /> renders as expected 1`] = `
<main
  className="add-dog-container"
>
  <h1
    className="form-title"
  >
    Add New Dog
  </h1>
  <form
    className="form-container"
    onSubmit={[Function]}
  >
    <div
      className="field-item"
    >
      <label
        className="bold"
        htmlFor="name"
      >
        Name
      </label>
      <input
        className="block"
        id="name"
        name="dogName"
        onChange={[Function]}
        required={true}
        type="text"
      />
    </div>
    <div
      className="field-item"
    >
      <label
        className="bold"
        htmlFor="image"
      >
        Image
      </label>
      <input
        accept="image/*"
        className="block"
        name="profileImg"
        onChange={[Function]}
        required={true}
        type="file"
      />
      <img
        alt="your-pic"
        className="img-preview"
        src=""
        style={
          Object {
            "display": "none",
          }
        }
      />
    </div>
    <fieldset
      className="field-item"
    >
      <legend
        className="bold"
      >
        Gender
      </legend>
      <label
        htmlFor="male"
      >
        <input
          id="male"
          name="gender"
          onChange={[Function]}
          required={true}
          type="radio"
          value="Male"
        />
        Male
      </label>
      <label
        htmlFor="female"
      >
        <input
          id="female"
          name="gender"
          onChange={[Function]}
          type="radio"
          value="Female"
        />
        Female
      </label>
    </fieldset>
    <label
      className="bold"
      htmlFor="estimated-age"
    >
      Estimated Birthdate
    </label>
    <div
      className="react-datepicker-wrapper"
    >
      <div
        className="react-datepicker__input-container"
      >
        <input
          className="block"
          disabled={false}
          id="age"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Click to select a date"
          readOnly={false}
          type="text"
          value=""
        />
      </div>
    </div>
    <label
      className="bold"
      htmlFor="arrival"
    >
      Estimated Arrival
    </label>
    <div
      className="react-datepicker-wrapper"
    >
      <div
        className="react-datepicker__input-container"
      >
        <input
          className="block"
          disabled={false}
          id="arrival"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Click to select a date"
          readOnly={false}
          type="text"
          value=""
        />
      </div>
    </div>
    <label
      className="bold"
      htmlFor="tag-number"
    >
      Tag Number
    </label>
    <input
      className="block"
      id="tag-number"
      name="tagNumber"
      onChange={[Function]}
      type="text"
    />
    <label
      className="bold"
      htmlFor="microchip"
    >
      Microchip Number
    </label>
    <input
      className="block"
      id="microchip"
      name="microchip"
      onChange={[Function]}
      type="text"
    />
    <fieldset
      className="field-item"
    >
      <legend
        className="bold"
      >
        Spayed/Neutered 
      </legend>
      <label
        htmlFor="yes"
      >
        <input
          id="yes"
          name="spayedNeutered"
          onChange={[Function]}
          required={true}
          type="radio"
        />
        Yes
      </label>
      <label
        htmlFor="no"
      >
        <input
          id="no"
          name="spayedNeutered"
          onChange={[Function]}
          type="radio"
        />
        No
      </label>
    </fieldset>
    <fieldset
      className="field-item "
    >
      <legend
        className="bold"
      >
        Required Shots Completed
      </legend>
      <label
        htmlFor="rabies"
      >
        Rabies
         
        <div
          className="react-datepicker-wrapper"
        >
          <div
            className="react-datepicker__input-container"
          >
            <input
              className="shot-date-input"
              disabled={false}
              name="rabies"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="Click to select a date"
              readOnly={false}
              type="text"
              value=""
            />
          </div>
        </div>
      </label>
      <label
        htmlFor="rabiesBooster"
      >
        Rabies Yearly Booster
         
        <div
          className="react-datepicker-wrapper"
        >
          <div
            className="react-datepicker__input-container"
          >
            <input
              className="shot-date-input"
              disabled={false}
              name="rabiesBooster"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="Click to select a date"
              readOnly={false}
              type="text"
              value=""
            />
          </div>
        </div>
      </label>
      <label
        htmlFor="complexOne"
      >
        Complex I
        <div
          className="react-datepicker-wrapper"
        >
          <div
            className="react-datepicker__input-container"
          >
            <input
              className="shot-date-input"
              disabled={false}
              name="complexOne"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="Click to select a date"
              readOnly={false}
              type="text"
              value=""
            />
          </div>
        </div>
      </label>
      <label
        htmlFor="complexTwo"
      >
        Complex II
        <div
          className="react-datepicker-wrapper"
        >
          <div
            className="react-datepicker__input-container"
          >
            <input
              className="shot-date-input"
              disabled={false}
              name="complexTwo"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="Click to select a date"
              readOnly={false}
              type="text"
              value=""
            />
          </div>
        </div>
      </label>
      <label
        htmlFor="complexBooster"
      >
        Complex Yearly Booster
        <div
          className="react-datepicker-wrapper"
        >
          <div
            className="react-datepicker__input-container"
          >
            <input
              className="shot-date-input"
              disabled={false}
              name="complexBooster"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="Click to select a date"
              readOnly={false}
              type="text"
              value=""
            />
          </div>
        </div>
      </label>
    </fieldset>
    <div
      className="form-buttons"
    >
      <button
        className="cancel"
      >
        <a
          className="dog-link"
          href="/dogs-list"
          onClick={[Function]}
        >
          Cancel
        </a>
      </button>
      <button
        className="submit"
        disabled="Name should have more than one letter."
        onClick={[Function]}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</main>
`;
